src/ser.rs

```diff
@@ -0,0 +1,183 @@
+use std::collections::HashMap;
+use std::io::{Read, Write, BufReader, Result, BufRead};
+use itertools::Itertools;
+use crate::{EGraph, Analysis, Language, Id, EClass, SymbolLang, ColorId};
+
+/// A trait for EGraphs that can be serialized.
+pub trait Serialization {
+    /// Exports graph data as a set of tuples. The format is `(op, eclass, children...)`.
+    fn to_tuples(&self) -> Vec<(String, Vec<Id>)>;
+    /// Exports graph data as a set of tuples. The format is one tuple per line,
+    /// space-separated: `op eclass children`.
+    fn to_tuples_text(&self, palette: &ColorPalette, out: &mut impl Write) -> Result<()>;
+}
+
+/// A trait for deserializing EGraphs from previously serialized content.
+pub trait Deserialization {
+    /// Constructs a graph from tuple data (eclass ids as `Id`).
+    fn from_tuples<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<Id>)>) -> (Self, ColorPalette) where Self: Sized;
+    /// Constructs a graph from tuple data (eclass ids as numbers).
+    fn from_tuples_int<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<usize>)>) -> (Self, ColorPalette) where Self: Sized;
+    /// Parses an constructs a graph from textual representation.
+    fn from_tuples_text(in_: &mut impl Read) -> Result<(Self, ColorPalette)> where Self: Sized;
+}
+
+trait DeserializationHelper<L: Language> {
+    fn add_class<'a>(&'a mut self, id: Id);
+    fn add_node(&mut self, eclass: Id, enode: L) -> &L;
+    fn update_parents(&mut self, parent: Id, enode: &L);
+    fn line_to_tuple(line: &str) -> (String, Vec<Id>);
+}
+
+pub struct ColorPalette {
+    colors: HashMap<Id, ColorId>
+}
+
+impl ColorPalette {
+    fn new() -> Self {
+        ColorPalette { colors: HashMap::default() }
+    }
+
+    fn get<L: Language, N: Analysis<L>>(&mut self, g: &mut EGraph<L, N>, color: Id) -> ColorId {
+        match self.colors.get(&color) {
+           Some(cid) => *cid,
+           None => {
+               let v = g.create_color();
+               self.colors.insert(color, v);
+               v
+           }
+        }
+    }
+}
+
+impl Default for ColorPalette {
+    fn default() -> Self { ColorPalette::new() }
+}
+
+impl<N: Analysis<SymbolLang>> Serialization for EGraph<SymbolLang, N> {
```

Comment:
We do not use Analysis<SymbolLang>, usualy N would just be ().
Also this impl should be in the EGraph file to prevent the pub(crate)